balancer/rls/picker.go

```diff
@@ -84,10 +84,8 @@ func (p *rlsPicker) Pick(info balancer.PickInfo) (balancer.PickResult, error) {
 	md, _ := metadata.FromOutgoingContext(info.Ctx)
 	reqKeys := p.kbm.RLSKey(md, p.origEndpoint, info.FullMethodName)
 
-	// Grab a read-lock to perform a cache lookup. If it so happens that we need
-	// to write to the cache (if we have to send out an RLS request), we will
-	// release the read-lock and acquire a write-lock.
-	p.lb.cacheMu.RLock()
+	p.lb.cacheMu.Lock()
```

Comment:
Can we reduce the amount of time we're holding this at all?

Reply:
The following actions are happening under the lock in this method:
- read the cache entry:
  - this is non-blocking and should be fast.
- send an RLS request if required:
  - this is the only thing which *could* block. But the control channel implementation spawns a goroutine to do the actual work of sending the RLS rpc. So, this should also be fast.
- queue the pick or delegate to child policies (or default policy):
  - this should also be non-blocking and should be fast.

So, I think we are OK to hold the lock for the whole duration of this method. If we are blocking or taking too long in this method, we have other problems as well I guess (since lb policies are expected to be quick in `Pick()` and non block in there).

What do you think?

Reply:
Seems fine, I was just wondering because the `defer` makes it impossible to give it up early on one path vs. another, and ideally we hold this for as short a time as possible.